lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java

```diff
@@ -0,0 +1,169 @@
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Hosts the actual runtime implementations and delegates invocations to them.
+ */
+final class RuntimesHost {
+  private final Set<String> runtimeDefinitions;
+  private Map<String, Runtime> runtimes;
+  private final Injector originalInjector;
+  private String defaultRuntimeName;
+
+  @Inject
+  private RuntimesHost(final Injector injector,
+                       @Parameter(SerializedRuntimeDefinitions.class) final Set<String> runtimeDefinitions) {
+    if (runtimeDefinitions == null || runtimeDefinitions.size() == 0) {
+      throw new RuntimeException("No runtime configurations are provided for multi-runtime");
+    }
+
+    this.runtimeDefinitions = runtimeDefinitions;
+    this.originalInjector = injector;
+  }
+
+  private static RuntimeDefinition parseSerializedRuntimeDefinition(final String serializedRuntimeDefinition) throws
+          IOException {
+    final RuntimeDefinition rd;
+    final JsonDecoder decoder = DecoderFactory.get().
+            jsonDecoder(RuntimeDefinition.getClassSchema(), serializedRuntimeDefinition);
+    final SpecificDatumReader<RuntimeDefinition> reader = new SpecificDatumReader<>(RuntimeDefinition.class);
+    rd = reader.read(null, decoder);
+    return rd;
+  }
+
+  private synchronized void initialize() {
+    if (this.runtimes != null) {
+      return;
+    }
+
+    this.runtimes = new HashMap<>();
+    final AvroConfigurationSerializer serializer = new AvroConfigurationSerializer();
+
+    for (final String serializedRuntimeDefinition : runtimeDefinitions) {
+      Configuration config = null;
+      RuntimeDefinition rd = null;
+      try {
+        rd = parseSerializedRuntimeDefinition(serializedRuntimeDefinition);
+
+        if (rd.getDefaultConfiguration()) {
+          this.defaultRuntimeName = rd.getRuntimeName().toString();
+        }
+
+        config = serializer.fromString(rd.getSerializedConfiguration().toString());
+      } catch (IOException e) {
+        throw new RuntimeException("Unable to read runtime configuarion.", e);
+      }
+
+      final Injector rootInjector = Tang.Factory.getTang().newInjector();
+      try {
+        initializeInjector(rootInjector);
+        final JavaConfigurationBuilder cb = Tang.Factory.getTang().newConfigurationBuilder();
+        cb.bindNamedParameter(RuntimeName.class, rd.getRuntimeName().toString());
+        cb.bindImplementation(Runtime.class, RuntimeImpl.class);
+
+        final Injector runtimeInjector = rootInjector.forkInjector(config, cb.build());
+        this.runtimes.put(rd.getRuntimeName().toString(), runtimeInjector.getInstance(Runtime.class));
+      } catch (InjectionException e) {
+        throw new RuntimeException("Unable to initialize runtimes.", e);
+      }
+    }
+  }
+
+  private void initializeInjector(final Injector runtimeInjector) throws InjectionException {
```

Comment:
I need to create different injector for each runtime as they define conflicting bindings. Also I cannot fork the original injector because of the same reason. What I do I create new injectors and copy form the original injector what I need.